File: automation/eimer_etikette.py
"""Eimer-Etikette fuer Massen-Schreiber (22.09.2026).

GEMESSEN 22.09. 20:27 UTC: Eimer 95 von 2'000 Punkten, waehrend vier Dauerlaeufer schrieben
(produkttexte_du_form, fortura_image_backfill, cj_reviews_import, cj_video_reel_engine). Zwei
Tages-Waechter starben in derselben Stunde mit «12x gedrosselt (Eimer dauerhaft leer)»:
cj_versand_ch_guard (Ghost-Sale-Klasse) und lagerstand_hygiene. Jeder Schreiber war fuer sich
hoeflich (schlief bei THROTTLED) — zusammen liefen sie den Eimer auf null und liessen ihn dort.
Die Schranke begrenzt, wie viele STARTEN; sie begrenzt nicht, wie viel ein Laufender TRINKT.

Regel: Wer in Masse schreibt, laesst dem Eimer einen Boden. Nach JEDER Antwort den
throttleStatus lesen; liegt currentlyAvailable unter BODEN, warten, bis ZIEL wieder erreicht ist
(aus restoreRate gerechnet, hoechstens DECKEL Sekunden). Tages-Waechter (leicht, kurz) brauchen
das nicht — sie sollen den Boden VORFINDEN.

Nutzung:  from eimer_etikette import nachlauf ; d = <Antwort-JSON> ; nachlauf(d)
Umgebung: EIMER_BODEN (Standard 600), EIMER_ZIEL (1000), EIMER_DECKEL (20 s), EIMER_AUS=1 schaltet ab.
"""
import os, time

BODEN = float(os.environ.get("EIMER_BODEN", "600"))
ZIEL = float(os.environ.get("EIMER_ZIEL", "1000"))
DECKEL = float(os.environ.get("EIMER_DECKEL", "20"))
AUS = os.environ.get("EIMER_AUS") == "1"
_gewartet = {"n": 0, "s": 0.0}


def nachlauf(d):
    """Liest throttleStatus aus einer GraphQL-Antwort; wartet, wenn der Eimer unter dem Boden liegt.
    Gibt die gewartete Zeit in Sekunden zurueck (0 = nichts zu tun)."""
    if AUS or not isinstance(d, dict):
        return 0.0
    ts = (((d.get("extensions") or {}).get("cost") or {}).get("throttleStatus") or {})
    try:
        avail = float(ts.get("currentlyAvailable")); rate = float(ts.get("restoreRate") or 0)
    except (TypeError, ValueError):
        return 0.0
    if avail >= BODEN or rate <= 0:
        return 0.0
    warte = min(DECKEL, max(1.0, (ZIEL - avail) / rate))
    _gewartet["n"] += 1; _gewartet["s"] += warte
    time.sleep(warte)
    return warte


def bilanz():
    return f"Eimer-Etikette: {_gewartet['n']}x gewartet, {_gewartet['s']:.0f} s gesamt"

File: automation/eimer_etikette.mjs
// Eimer-Etikette fuer Massen-Schreiber (Node) — Begruendung und Regel: siehe eimer_etikette.py.
// Nutzung:  import { nachlauf } from './eimer_etikette.mjs';  const j = await r.json(); await nachlauf(j);
const BODEN = parseFloat(process.env.EIMER_BODEN || '600');
const ZIEL = parseFloat(process.env.EIMER_ZIEL || '1000');
const DECKEL = parseFloat(process.env.EIMER_DECKEL || '20');
const AUS = process.env.EIMER_AUS === '1';
export const bilanz = { n: 0, s: 0 };
export async function nachlauf(j) {
  if (AUS || !j || typeof j !== 'object') return 0;
  const ts = j.extensions?.cost?.throttleStatus || {};
  const avail = Number(ts.currentlyAvailable), rate = Number(ts.restoreRate || 0);
  if (!Number.isFinite(avail) || avail >= BODEN || !(rate > 0)) return 0;
  const warte = Math.min(DECKEL, Math.max(1, (ZIEL - avail) / rate));
  bilanz.n++; bilanz.s += warte;
  await new Promise(r => setTimeout(r, warte * 1000));
  return warte;
}
